### PR TITLE
Refactor roleset into utils mk II

### PR DESCRIPTION
**Pull request by lawliet89**
## Overview

Hi @kalafut, this is a clean up of @emilymye's PR in https://github.com/hashicorp/vault-plugin-secrets-gcp/pull/62

I have done the following:

- Merge `master` in to fix conflicts
- Fix some integration tests

If you could get this merged in, I will then proceed to rebase https://github.com/hashicorp/vault-plugin-secrets-gcp/pull/67 and hopefully get that in too. I have a WIP rebase [here](https://github.com/lawliet89/vault-plugin-secrets-gcp/compare/roleset_refactor...lawliet89:static_account).

## Acceptance Tests Output

```
$ make test-acc
?       github.com/hashicorp/vault-plugin-secrets-gcp   [no test files]
=== RUN   TestConfigRotateRootUpdate
=== PAUSE TestConfigRotateRootUpdate
=== RUN   TestConfig
=== PAUSE TestConfig
=== RUN   TestPathRoleSet_Basic
--- PASS: TestPathRoleSet_Basic (19.09s)
=== RUN   TestPathRoleSet_UpdateKeyRoleSet
--- PASS: TestPathRoleSet_UpdateKeyRoleSet (28.79s)
=== RUN   TestPathRoleSet_RotateKeyRoleSet
--- PASS: TestPathRoleSet_RotateKeyRoleSet (27.80s)
=== RUN   TestPathRoleSet_UpdateTokenRoleSetScopes
--- PASS: TestPathRoleSet_UpdateTokenRoleSetScopes (22.50s)
=== RUN   TestPathRoleSet_UpdateTokenRoleSet
--- PASS: TestPathRoleSet_UpdateTokenRoleSet (34.81s)
=== RUN   TestPathRoleSet_RotateTokenRoleSet
--- PASS: TestPathRoleSet_RotateTokenRoleSet (32.50s)
=== RUN   TestSecrets_GenerateAccessToken
--- PASS: TestSecrets_GenerateAccessToken (15.56s)
=== RUN   TestSecrets_GenerateKeyConfigTTL
2020/11/18 16:39:47 [DEBUG] test check failed with error Get "https://iam.googleapis.com/v1/projects/yw-testing/roles?alt=json&prettyPrint=false": oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid JWT Signature."} (attempt 0), sleeping one second before trying again
--- PASS: TestSecrets_GenerateKeyConfigTTL (19.93s)
=== RUN   TestSecrets_GenerateKeyTTLOverride
2020/11/18 16:40:06 [DEBUG] test check failed with error Get "https://iam.googleapis.com/v1/projects/yw-testing/roles?alt=json&prettyPrint=false": oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid JWT Signature."} (attempt 0), sleeping one second before trying again
--- PASS: TestSecrets_GenerateKeyTTLOverride (16.77s)
=== RUN   TestSecrets_GenerateKeyMaxTTLCheck
2020/11/18 16:40:24 [DEBUG] test check failed with error Get "https://iam.googleapis.com/v1/projects/yw-testing/roles?alt=json&prettyPrint=false": oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid JWT Signature."} (attempt 0), sleeping one second before trying again
--- PASS: TestSecrets_GenerateKeyMaxTTLCheck (17.63s)
=== CONT  TestConfigRotateRootUpdate
=== CONT  TestConfig
=== RUN   TestConfigRotateRootUpdate/no_configuration
=== PAUSE TestConfigRotateRootUpdate/no_configuration
=== RUN   TestConfigRotateRootUpdate/config_with_no_credentials
=== PAUSE TestConfigRotateRootUpdate/config_with_no_credentials
=== RUN   TestConfigRotateRootUpdate/config_with_invalid_credentials
=== PAUSE TestConfigRotateRootUpdate/config_with_invalid_credentials
=== RUN   TestConfigRotateRootUpdate/rotate
=== PAUSE TestConfigRotateRootUpdate/rotate
--- PASS: TestConfig (0.00s)
=== CONT  TestConfigRotateRootUpdate/no_configuration
=== CONT  TestConfigRotateRootUpdate/rotate
=== CONT  TestConfigRotateRootUpdate/config_with_invalid_credentials
=== CONT  TestConfigRotateRootUpdate/config_with_no_credentials
2020/11/18 16:40:38 [DEBUG] test check failed with error failed to create new key: Post "https://iam.googleapis.com/v1/projects/-/serviceAccounts/vault-tester@yw-testing.iam.gserviceaccount.com/keys?alt=json&prettyPrint=false": oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid JWT Signature."} (attempt 0), sleeping one second before trying again
=== CONT  TestConfigRotateRootUpdate/rotate
    path_config_rotate_root_test.go:201: WARNING: failed to delete key created for test, clean up manually: googleapi: Error 400: Precondition check failed., failedPrecondition
--- PASS: TestConfigRotateRootUpdate (0.00s)
    --- PASS: TestConfigRotateRootUpdate/no_configuration (0.00s)
    --- PASS: TestConfigRotateRootUpdate/config_with_invalid_credentials (0.00s)
    --- PASS: TestConfigRotateRootUpdate/config_with_no_credentials (0.00s)
    --- PASS: TestConfigRotateRootUpdate/rotate (8.17s)
PASS
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin    243.572s
=== RUN   TestIamResource_ServiceAccount
--- PASS: TestIamResource_ServiceAccount (7.28s)
=== RUN   TestPolicyToDataset
--- PASS: TestPolicyToDataset (0.00s)
=== RUN   TestDatasetToPolicy
--- PASS: TestDatasetToPolicy (0.00s)
=== RUN   TestDatasetResource
--- PASS: TestDatasetResource (0.00s)
=== RUN   TestConditionalDatasetResource
--- PASS: TestConditionalDatasetResource (0.00s)
=== RUN   TestIamResource
--- PASS: TestIamResource (0.00s)
=== RUN   TestConditionalIamResource
--- PASS: TestConditionalIamResource (0.00s)
=== RUN   TestEnabledIamResources_RelativeName
--- PASS: TestEnabledIamResources_RelativeName (0.01s)
=== RUN   TestEnabledIamResources_FullName
--- PASS: TestEnabledIamResources_FullName (0.00s)
=== RUN   TestEnabledIamResources_SelfLink
--- PASS: TestEnabledIamResources_SelfLink (0.00s)
=== RUN   TestIamEnabledResources_ValidateGeneratedConfig
--- PASS: TestIamEnabledResources_ValidateGeneratedConfig (0.00s)
PASS
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil    7.300s
?       github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil/internal   [no test files]
=== RUN   TestParseBindings
--- PASS: TestParseBindings (0.00s)
=== RUN   TestParseBindingsB64
--- PASS: TestParseBindingsB64 (0.00s)
PASS
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin/util       (cached)
?       github.com/hashicorp/vault-plugin-secrets-gcp/scripts/gohelpers [no test files]
```

## Contributor Checklist

- [x] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet (No Docs)
- [x] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
- [x] Backwards compatible

### CODE REVIEW

**Comment by lawliet89** [comment]
```
$ make test-acc
?       github.com/hashicorp/vault-plugin-secrets-gcp   [no test files]
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin    399.185s
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil    3.171s
?       github.com/hashicorp/vault-plugin-secrets-gcp/plugin/iamutil/internal   [no test files]
ok      github.com/hashicorp/vault-plugin-secrets-gcp/plugin/util       0.005s
?       github.com/hashicorp/vault-plugin-secrets-gcp/scripts/gohelpers [no test files]
```

**Comment by lawliet89** [comment]
Hey @kalafut, could this be reviewed ASAP before more conflicts creep in?

**Comment by kalafut** [comment]
Yeah. There are some holidays at the moment but I don’t expect anything to go in ahead of this (ie no more conflicts). Thanks!

**Comment by lawliet89** [comment]
Sorry I forgot about the Thanksgiving holidays. Have a great time :)

**Comment by lawliet89** [comment]
Hey @kalafut sorry to prod you again. Could this be looked at please?

File: plugin/backend_test.go [hunk 1]
```diff
@@ -2,10 +2,12 @@ package gcpsecrets
 
 import (
 	"context"
+	"os"
 	"testing"
 	"time"
 
-	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-hclog"
+
```
Comment (kalafut) [inline plugin/backend_test.go:10]:
```suggestion
```

File: plugin/backend_test.go [hunk 2]
```diff
@@ -14,12 +16,14 @@ const (
...
 )
 
-func getTestBackend(tb testing.TB) (logical.Backend, logical.Storage) {
+func getTestBackend(tb testing.TB) (*backend, logical.Storage) {
 	tb.Helper()
 
 	config := logical.TestBackendConfig()
 	config.StorageView = new(logical.InmemStorage)
-	config.Logger = hclog.NewNullLogger()
+	if os.Getenv("VAULT_LOG") == "" {
```
Comment (kalafut) [inline plugin/backend_test.go:24]: What does this do? I'm assuming if `config.Logger == nil` other stuff might have issues?

File: plugin/role_set.go [hunk 3]
```diff
@@ -93,319 +106,259 @@ func (rs *RoleSet) bindingHash() string {
...
+	newWalIds, err := b.addWalsForRoleSetResources(ctx, req, rs.Name, newResources)
 	if err != nil {
-		tryDeleteWALs(ctx, s, oldWals...)
-		return nil, errwrap.Wrapf("failed to create WAL for cleaning up old account: {{err}}", err)
+		return nil, err
 	}
 
-	newWals := make([]string, 0, len(newBinds)+2)
-	walId, err := rs.newServiceAccount(ctx, s, iamAdmin, project)
+	// Created new RoleSet resources
```
Comment (kalafut) [inline plugin/role_set.go:164]: What is this referring to?

File: plugin/rollback.go [hunk 4]
```diff
@@ -4,12 +4,13 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/hashicorp/vault/sdk/framework"
```
Comment (kalafut) [inline plugin/rollback.go:7]: Rerun `goimports`

File: plugin/secrets_service_account_key.go [hunk 5]
```diff
@@ -175,25 +175,17 @@ func (b *backend) secretKeyRevoke(ctx context.Context, req *logical.Request, d *
 
 func (b *backend) getSecretKey(ctx context.Context, s logical.Storage, rs *RoleSet, keyType, keyAlgorithm string, ttl int) (*logical.Response, error) {
 	cfg, err := getConfig(ctx, s)
```
Comment (kalafut) [inline plugin/secrets_service_account_key.go:177]: This error isn't checked.

File: plugin/secrets_service_account_key.go [hunk 6]
```diff
@@ -175,25 +175,17 @@ func (b *backend) secretKeyRevoke(ctx context.Context, req *logical.Request, d *
...
-	if err != nil {
-		return nil, errwrap.Wrapf("could not read backend config: {{err}}", err)
-	}
-	if cfg == nil {
-		cfg = &config{}
+	if rs.AccountId == nil {
```
Comment (calvn) [inline plugin/secrets_service_account_key.go:178]: We should still check for nil on `cfg` and do an `err` check before this.

File: plugin/rollback.go (hunk 4 above)
Comment (lawliet89) [reply plugin/rollback.go]: Done

File: plugin/rollback.go [hunk 7]
```diff
@@ -105,8 +106,8 @@ func (b *backend) serviceAccountKeyRollback(ctx context.Context, req *logical.Re
 
 	// If roleset is not nil, get key in use.
 	if rs != nil {
-		if rs.SecretType == SecretTypeAccessToken {
-			// Don't clean keys if roleset generates key secrets.
+		if rs.SecretType != SecretTypeAccessToken {
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply plugin/rollback.go:108]: Looks like it, yes. Was originally in this commit: https://github.com/emilymye/vault-plugin-secrets-gcp/commit/6b276532d40ce8d73b727985f84bde384740def0

File: plugin/secrets_service_account_key.go (hunk 6 above)
Comment (lawliet89) [reply plugin/secrets_service_account_key.go:184]: Added back in. Looks like a mess up during the rebase/merge.

File: plugin/secrets_service_account_key.go (hunk 5 above)
Comment (lawliet89) [reply plugin/secrets_service_account_key.go]: Added back in. Looks like a mess up during the rebase/merge.

File: plugin/backend_test.go (hunk 2 above)
Comment (lawliet89) [reply plugin/backend_test.go]: Made changes in f16c3d6ec7dbe88b00eacbb6320a0675743b2363

File: plugin/role_set.go (hunk 3 above)
Comment (lawliet89) [reply plugin/role_set.go:164]: This was copied verbatim from #62.

I am guessing this is referring to the bunch of GCP resources that will have to be created for this roleset, including:

- GCP Service Account
- IAM Bindings
- Service Account key if this was an `access_token`.

**Review by calvn — APPROVED** [review]

**Comment by lawliet89** [comment]
Thanks for the approval @calvn! Once this is merged in, I'll rebase and open "part two" of this PR to support GCP static accounts that Emily previously worked on.

**Comment by calvn** [comment]
@lawliet89 got a couple of question while re-visiting this once more before merging:

1. Are the debug output failures from `TestSecrets_GenerateKey*` expected?
1. `TestConfigRotateRootUpdate/rotate` seems to be outputting a warning during its cleanup func. Not sure if it's [related to this](https://github.com/googleapis/google-api-go-client/issues/645), but it may look misleading/error-checking improperly if it's always outputting that error. Is this something we can safely ignore?

**Comment by lawliet89** [comment]
@calvn the errors you see are because GCP IAM is eventually consistent. For example, a service account creation/deletion might not be reflected immediately. 

> Are the debug output on retry failures from TestSecrets_GenerateKey* expected?

Are these the outputs you are seeing? If so, then it is related to the eventual consistency aspects of GCP IAM. The tests have the auto-retry built in for this reason.

```
2021/02/17 10:09:45 [DEBUG] test check failed with error Get "https://iam.googleapis.com/v1/projects/yw-testing/roles?alt=json&prettyPrint=false": oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid JWT Signature."} (attempt 0), sleeping one second before trying again
```

> TestConfigRotateRootUpdate/rotate seems to be outputting a warning during its cleanup func. Not sure if it's related to this, but it may look misleading/error-checking improperly if it's always outputting that error. Is this something we can safely ignore?

Same thing. I couldn't capture the debug output on my run but I do seem them from time to time.

**Comment by calvn** [comment]
My observation was from the test output that you posted originally. The debug/warning messages were present even before this changeset right? Just double checking to make sure that we're not changing behavior here.

**Comment by lawliet89** [comment]
@calvn yes they're not new.

**Comment by lawliet89** [comment]
Thanks for the review and merging in @calvn and @kalafut 

I've opened https://github.com/hashicorp/vault-plugin-secrets-gcp/pull/107 which is the second part of this PR.

Thanks!

**Pull request merged**